<launch>
  <!-- Arguments -->
  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>
  <arg name="map_file" default="$(find multinav_turtlebot3)/maps/emptyHouse_map/empty_house.yaml"/>
  <arg name="open_rviz" default="false"/>
  <arg name="move_forward_only" default="false"/>
  <arg name="first_tb3" default="tb3_0" />
  <arg name="second_tb3" default="tb3_1" />
  <arg name="third_tb3" default="tb3_2" />
  <arg name="fourth_tb3" default="tb3_3" />
  <arg name="use_map_topic"   default="false"/>

  <arg name="scan_topic"      default="/tb3_1/scan"/>
  <arg name="odom_frame_id"   default="tb3_1/odom"/>
  <arg name="base_frame_id"   default="tb3_1/base_footprint"/>
  <arg name="global_frame_id" default="map"/>

  <arg name="initial_pose_x" default="-1.0"/>
  <arg name="initial_pose_y" default="1.0"/>
  <arg name="initial_pose_a" default="0.0"/>

  <!-- AMCL -->
 <node pkg="amcl" type="amcl" name="tb3_1_amcl">

   <param name="use_map_topic"             value="$(arg use_map_topic)"/>
   <param name="initial_pose_x"            value="$(arg initial_pose_x)"/>
   <param name="initial_pose_y"            value="$(arg initial_pose_y)"/>
   <param name="initial_pose_a"            value="$(arg initial_pose_a)"/>

   <param name="min_particles"             value="500"/>
   <param name="max_particles"             value="3000"/>
   <param name="kld_err"                   value="0.02"/>
   <param name="update_min_d"              value="0.20"/>
   <param name="update_min_a"              value="0.20"/>
   <param name="resample_interval"         value="1"/>
   <param name="transform_tolerance"       value="0.5"/>
   <param name="recovery_alpha_slow"       value="0.00"/>
   <param name="recovery_alpha_fast"       value="0.00"/>
   <param name="initial_pose_x"            value="-1.00"/>
   <param name="initial_pose_y"            value="1.00"/>
   <param name="initial_pose_a"            value="0.00"/>
   <param name="gui_publish_rate"          value="50.0"/>


   <param name="laser_max_range"           value="3.5"/>
   <param name="laser_max_beams"           value="180"/>
   <param name="laser_z_hit"               value="0.5"/>
   <param name="laser_z_short"             value="0.05"/>
   <param name="laser_z_max"               value="0.05"/>
   <param name="laser_z_rand"              value="0.5"/>
   <param name="laser_sigma_hit"           value="0.2"/>
   <param name="laser_lambda_short"        value="0.1"/>
   <param name="laser_likelihood_max_dist" value="2.0"/>
   <param name="laser_model_type"          value="likelihood_field"/>

   <param name="odom_model_type"           value="diff"/>
   <param name="odom_alpha1"               value="0.1"/>
   <param name="odom_alpha2"               value="0.1"/>
   <param name="odom_alpha3"               value="0.1"/>
   <param name="odom_alpha4"               value="0.1"/>
   <param name="odom_frame_id"             value="$(arg odom_frame_id)"/>
   <param name="base_frame_id"             value="$(arg base_frame_id)"/>
   <param name="global_frame_id"           value="$(arg global_frame_id)"/>

   <remap from="scan"                      to="/tb3_1/scan"/>
   <remap from="initialpose"               to="/tb3_1/initialpose"/>
   <remap from="amcl_pose"                 to="/tb3_1/amcl_pose"/>
   <remap from="particlecloud"             to="/tb3_1/particlecloud"/>

 </node>


  <!-- move_base -->
  <!-- Arguments -->
  <arg name="odom_topic" default="/tb3_1/odom" />
  <arg name="laser_topic" default="/tb3_1/laser/scan" />

  <!-- move_base -->
  <node pkg="move_base" type="move_base" respawn="false" name="move_base2" output="screen">
    <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS" />

    <rosparam file="$(find multinav_turtlebot3)/param/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find multinav_turtlebot3)/param/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find multinav_turtlebot3)/param/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find multinav_turtlebot3)/param/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find multinav_turtlebot3)/param/dwa_local_planner_params.yaml" command="load" />
    <rosparam file="$(find multinav_turtlebot3)/param/move_base_params.yaml" command="load" />
    <param name="DWAPlannerROS/min_vel_x" value="0.0" if="$(arg move_forward_only)" />

    <param name="global_costmap/global_frame" value="$(arg global_frame_id)"/>
    <param name="global_costmap/robot_base_frame" value="$(arg base_frame_id)"/>
    <param name="local_costmap/global_frame" value="$(arg odom_frame_id)"/>
    <param name="local_costmap/robot_base_frame" value="$(arg base_frame_id)"/>
    <param name="DWAPlannerROS/global_frame_id" value="$(arg odom_frame_id)"/>

    <remap from="cmd_vel" to="/tb3_1/cmd_vel"/>
    <remap from="odom" to="$(arg odom_topic)"/>
    <remap from="scan" to="$(arg laser_topic)"/>
    <remap from="map" to="/map" />
    <remap from="/move_base_simple/goal"                                           to="/tb3_1/move_base_simple/goal" />
    <remap from="/move_base/TebLocalPlannerROS/global_plan"                        to="/tb3_1/move_base/TebLocalPlannerROS/global_plan" />
    <remap from="/move_base/TebLocalPlannerROS/local_plan"                         to="/tb3_1/move_base/TebLocalPlannerROS/local_plan" />
    <remap from="/move_base/TebLocalPlannerROS/teb_markers"                        to="/tb3_1/move_base/TebLocalPlannerROS/teb_markers" />
    <remap from="/move_base/TebLocalPlannerROS/teb_markers_array"                  to="/tb3_1/move_base/TebLocalPlannerROS/teb_markers_array" />
    <remap from="/move_base/TebLocalPlannerROS/teb_poses"                          to="/tb3_1/move_base/TebLocalPlannerROS/teb_poses" />
    <remap from="/move_base/global_costmap/costmap"                                to="/tb3_1/move_base/global_costmap/costmap" />
    <remap from="/move_base/global_costmap/costmap_updates"                        to="/tb3_1/move_base/global_costmap/costmap_updates" />
    <remap from="/move_base/local_costmap/costmap"                                 to="/tb3_1/move_base/local_costmap/costmap" />
    <remap from="/move_base/local_costmap/costmap_updates"                         to="/tb3_1/move_base/local_costmap/costmap_updates" />
    <remap from="/move_base/local_costmap/footprint"                               to="/tb3_1/move_base/local_costmap/footprint" />

    <remap from="/move_base/GlobalPlanner/parameter_descriptions"                  to="/tb3_1/move_base/GlobalPlanner/parameter_descriptions" />
    <remap from="/move_base/GlobalPlanner/parameter_updates"                       to="/tb3_1/move_base/GlobalPlanner/parameter_updates" />
    <remap from="/move_base/GlobalPlanner/plan"                                    to="/tb3_1/move_base/GlobalPlanner/plan" />
    <remap from="/move_base/GlobalPlanner/potential"                               to="/tb3_1/move_base/GlobalPlanner/potential" />
    <remap from="/move_base/TebLocalPlannerROS/obstacles"                          to="/tb3_1/move_base/TebLocalPlannerROS/obstacles" />
    <remap from="/move_base/TebLocalPlannerROS/parameter_descriptions"             to="/tb3_1/move_base/TebLocalPlannerROS/parameter_descriptions" />
    <remap from="/move_base/TebLocalPlannerROS/parameter_updates"                  to="/tb3_1/move_base/TebLocalPlannerROS/parameter_updates" />
    <remap from="/move_base/cancel"                                                to="/tb3_1/move_base/cancel" />
    <remap from="/move_base/current_goal"                                          to="/tb3_1/move_base/current_goal" />
    <remap from="/move_base/feedback"                                              to="/tb3_1/move_base/feedback" />
    <remap from="/move_base/global_costmap/footprint"                              to="/tb3_1/move_base/global_costmap/footprint" />
    <remap from="/move_base/global_costmap/inflation_layer/parameter_descriptions" to="/tb3_1/move_base/global_costmap/inflation_layer/parameter_descriptions" />
    <remap from="/move_base/global_costmap/inflation_layer/parameter_updates"      to="/tb3_1/move_base/global_costmap/inflation_layer/parameter_updates" />
    <remap from="/move_base/global_costmap/obstacle_layer/clearing_endpoints"      to="/tb3_1/move_base/global_costmap/obstacle_layer/clearing_endpoints" />
    <remap from="/move_base/global_costmap/obstacle_layer/parameter_descriptions"  to="/tb3_1/move_base/global_costmap/obstacle_layer/parameter_descriptions" />
    <remap from="/move_base/global_costmap/obstacle_layer/parameter_updates"       to="/tb3_1/move_base/global_costmap/obstacle_layer/parameter_updates" />
    <remap from="/move_base/global_costmap/parameter_descriptions"                 to="/tb3_1/move_base/global_costmap/parameter_descriptions" />
    <remap from="/move_base/global_costmap/parameter_updates"                      to="/tb3_1/move_base/global_costmap/parameter_updates" />
    <remap from="/move_base/global_costmap/static_layer/parameter_descriptions"    to="/tb3_1/move_base/global_costmap/static_layer/parameter_descriptions" />
    <remap from="/move_base/global_costmap/static_layer/parameter_updates"         to="/tb3_1/move_base/global_costmap/static_layer/parameter_updates" />
    <remap from="/move_base/goal"                                                  to="/tb3_1/move_base/goal" />
    <remap from="/move_base/local_costmap/obstacle_layer/parameter_descriptions"   to="/tb3_1/move_base/local_costmap/obstacle_layer/parameter_descriptions" />
    <remap from="/move_base/local_costmap/obstacle_layer/parameter_updates"        to="/tb3_1/move_base/local_costmap/obstacle_layer/parameter_updates" />
    <remap from="/move_base/local_costmap/parameter_descriptions"                  to="/tb3_1/move_base/local_costmap/parameter_descriptions" />
    <remap from="/move_base/local_costmap/parameter_updates"                       to="/tb3_1/move_base/local_costmap/parameter_updates" />
    <remap from="/move_base/local_costmap/static_layer/parameter_descriptions"     to="/tb3_1/move_base/local_costmap/static_layer/parameter_descriptions" />
    <remap from="/move_base/local_costmap/static_layer/parameter_updates"          to="/tb3_1/move_base/local_costmap/static_layer/parameter_updates" />
    <remap from="/move_base/parameter_descriptions"                                to="/tb3_1/move_base/parameter_descriptions" />
    <remap from="/move_base/parameter_updates"                                     to="/tb3_1/move_base/parameter_updates" />
    <remap from="/move_base/result"                                                to="/tb3_1/move_base/result" />
    <remap from="/move_base/status"                                                to="/tb3_1/move_base/status" />
    <remap from="/move_base_simple/goal"                                           to="/tb3_1/move_base_simple/goal" />
  </node>


</launch>